### DISCUSSION TITLE
Usage with vanilla JS or other frameworks

### DISCUSSION THREAD

**Discussion by lehno** [umodoc/editor#64] [Ideas]
How hard would it be to have the editor in pure vanilla JS, so It could easily be used in other frameworks? 
We are using angular and we are aiming to have this editor running in angular but in a way it can also run without angular.
We have our own tiptap editor there, but we love the idea of pages and be able to page break and print docs

Thanks for any help

**Comment by umodoc** [comment]
Hello @lehno ,

Thank you for your interest in Umo Editor and for raising your question. From a technical standpoint, the difficulty of converting Umo Editor to pure vanilla JavaScript is not significant, but it does require a considerable amount of work. This includes rewriting the UI framework, Vue components, and managing global state, among other things. Given the numerous files involved in Umo Editor, this would be a massive and patient-testing task. At present, we have not considered implementing or refactoring Umo Editor entirely in pure vanilla JavaScript, but we are very open to the idea of you adopting parts of Umo Editor's source code in your own tiptap editor to implement features like pagination, page breaks, and document printing.

We are eagerly looking forward to any progress you might share. Here are some core code files that might be helpful to you:

- Page display: [Page.vue](https://github.com/umodoc/editor/blob/main/src/components/container/page.vue)
- Page printing: [Print.vue](https://github.com/umodoc/editor/blob/main/src/components/container/print.vue)
- Tiptap pagination extension, implemented by [Cassielxd](https://github.com/Cassielxd): [page extension](https://github.com/umodoc/editor/tree/main/src/extensions/page)

We hope this information is helpful to you. Thank you again for your appreciation of Umo Editor, and we hope you can make progress soon.

Best regards,
The Umo Team

---

感谢您对Umo Editor的兴趣以及您提出的问题。从技术角度来看，将Umo Editor转为纯原生JavaScript实现的难度并不大，但确实需要进行大量的工作。这包括重写UI框架、Vue组件以及管理全局状态等。由于Umo Editor涉及的文件众多，这将是一项庞大且需要耐心的任务。目前，我们尚未考虑将Umo Editor完全用纯原生JavaScript实现或重构，但我们非常欢迎您在自己的tiptap编辑器中采用Umo Editor的部分源码，以实现分页、页面打印等功能。

如果您有任何进展，我们非常期待您的分享。以下是一些可能对您有帮助的核心代码文件链接：

- 分页展示：[Page.vue](https://github.com/umodoc/editor/blob/main/src/components/container/page.vue)
- 页面打印：[Print.vue](https://github.com/umodoc/editor/blob/main/src/components/container/print.vue)
- tiptap分页扩展，由[Cassielxd](https://github.com/Cassielxd)实现：[page extension](https://github.com/umodoc/editor/tree/main/src/extensions/page)

希望这些信息对您有所帮助。再次感谢您对Umo Editor的喜爱，我们期待您能尽快取得进展。

**Reply by lehno** [reply]
I've sent an email 2 weeks ago regarding a quick chat and see if you are interested in some agreement for this option?

**Reply by umodoc** [reply]
> I've sent an email 2 weeks ago regarding a quick chat and see if you are interested in some agreement for this option?

Apologies for the late response, I've just noticed your email. If you have any ideas, feel free to reach out to me via email at any time.

**Comment by Goldziher** [comment]
I believe you wrap the editor as a web component without much effort. Vue excels in this. 

https://vuejs.org/guide/extras/web-components
